File: nRainhas.py
import time


def posicao_segura(linha, coluna, pos):
    for i in range(linha):
        # mesma coluna
        if pos[i] == coluna:
            return 0

        # mesma diagonal
        if abs(pos[i] - coluna) == abs(i - linha):
            return 0

    return 1


def resolver_n_rainhas(linha, pos, total_solucoes, n):
    if linha == n:
        total_solucoes[0] = total_solucoes[0] + 1
        return

    for coluna in range(n):
        if posicao_segura(linha, coluna, pos):
            pos[linha] = coluna
            resolver_n_rainhas(linha + 1, pos, total_solucoes, n)


def main():
    total_solucoes = [0]

    n = int(input())

    pos = [0] * n

    inicio = time.monotonic()

    total_solucoes[0] = 0
    resolver_n_rainhas(0, pos, total_solucoes, n)

    fim = time.monotonic()

    tempo = fim - inicio

    print(f"n = {n}, solucoes = {total_solucoes[0]}, tempo = {tempo:.6f} segundos")


if name == "main":
    main()

File: Java/NRainhas.java
import java.util.Scanner;

public class NRainhas {

    public static boolean posicaoSegura(int linha, int coluna, int[] pos) {
        for (int i = 0; i < linha; i++) {

            // mesma coluna
            if (pos[i] == coluna) {
                return false;
            }

            // mesma diagonal
            if (Math.abs(pos[i] - coluna) == Math.abs(i - linha)) {
                return false;
            }
        }
        return true;
    }

    public static void resolverNRainhas(
            int linha,
            int[] pos,
            int n,
            int[] totalSolucoes,
            int[] totalChamadasRNR) {

        totalChamadasRNR[0]++;

        if (linha == n) {
            totalSolucoes[0]++;
            return;
        }

        for (int coluna = 0; coluna < n; coluna++) {
            if (posicaoSegura(linha, coluna, pos)) {
                pos[linha] = coluna;
                resolverNRainhas(
                        linha + 1,
                        pos,
                        n,
                        totalSolucoes,
                        totalChamadasRNR);
            }
        }
    }

    public static void main(String[] args) {

        Scanner scanner = new Scanner(System.in);

        int n = scanner.nextInt();

        int[] pos = new int[n];

        int[] totalSolucoes = {0};
        int[] totalChamadasRNR = {0};

        long inicio = System.nanoTime();

        resolverNRainhas(
                0,
                pos,
                n,
                totalSolucoes,
                totalChamadasRNR);

        long fim = System.nanoTime();

        double tempo = (fim - inicio) / 1_000_000_000.0;

        System.out.printf(
                "n = %d, solucoes = %d, tempo = %.6f segundos, chamadasRNR = %d%n",
                n,
                totalSolucoes[0],
                tempo,
                totalChamadasRNR[0]);

        scanner.close();
    }
}
